pkg/sdkserver/sdkserver_test.go

```diff
@@ -678,176 +685,185 @@ func TestSDKServerUpdateEventHandler(t *testing.T) {
-		gs := defaultGs.DeepCopy()
-		gs.Status.State = v1alpha1.GameServerStateShutdown
-		_, sc, stop := setup(t, gs.DeepCopy())
-		defer close(stop)
-
-		_, err := sc.Allocate(context.Background(), &sdk.Empty{})
-		assert.EqualError(t, err, "cannot Allocate a Shutdown GameServer")
+	// wait for the game server to go back to being Ready
+	assertStateChange(v1alpha1.GameServerStateRequestReady, 4*time.Second, func(status v1alpha1.GameServerStatus) {
+		assert.Nil(t, status.ReservedUntil)
 	})
 
-	t.Run("Deleting gameserver", func(t *testing.T) {
-		gs := defaultGs.DeepCopy()
-		now := metav1.Now()
-		gs.ObjectMeta.DeletionTimestamp = &now
-		_, sc, stop := setup(t, gs.DeepCopy())
-		defer close(stop)
+	// now test for being able to escape the Reserved state when requesting another state
+
+	// test ready
+	_, err = sc.Reserve(context.Background(), &sdk.Duration{Seconds: 3})
+	assert.NoError(t, err)
+	assertStateChange(v1alpha1.GameServerStateReserved, 2*time.Second, noAdditional)
 
-		_, err := sc.Allocate(context.Background(), &sdk.Empty{})
-		assert.NoError(t, err)
+	_, err = sc.Ready(context.Background(), &sdk.Empty{})
+	assert.NoError(t, err)
+	assertStateChange(v1alpha1.GameServerStateRequestReady, 2*time.Second, func(status v1alpha1.GameServerStatus) {
+		assert.Nil(t, status.ReservedUntil)
 	})
 
-	t.Run("Unexpected error", func(t *testing.T) {
-		m, sc, stop := setup(t, defaultGs.DeepCopy())
-		defer close(stop)
-		done := make(chan struct{}, 10)
+	select {
+	case current := <-state:
+		assert.Failf(t, "Should not get update:", string(current.State))
+	case <-time.After(4 * time.Second):
+	}
 
-		m.AgonesClient.AddReactor("update", "gameservers", func(action k8stesting.Action) (bool, runtime.Object, error) {
-			defer func() {
-				done <- struct{}{}
-			}()
+	// test allocated
+	_, err = sc.Reserve(context.Background(), &sdk.Duration{Seconds: 3})
+	assert.NoError(t, err)
+	assertStateChange(v1alpha1.GameServerStateReserved, 2*time.Second, noAdditional)
 
-			return true, nil, errors.New("a bad thing happened")
-		})
+	_, err = sc.Allocate(context.Background(), &sdk.Empty{})
+	assert.NoError(t, err)
+	assertStateChange(v1alpha1.GameServerStateAllocated, 2*time.Second, func(status v1alpha1.GameServerStatus) {
+		assert.Nil(t, status.ReservedUntil)
+	})
+
+	select {
+	case <-state:
+		assert.Fail(t, "Should not go back to RequestReady")
+	case <-time.After(4 * time.Second):
+	}
+
+	// test shutdown
+	_, err = sc.Reserve(context.Background(), &sdk.Duration{Seconds: 3})
+	assert.NoError(t, err)
+	assertStateChange(v1alpha1.GameServerStateReserved, 2*time.Second, noAdditional)
 
-		_, err := sc.Allocate(context.Background(), &sdk.Empty{})
-		assert.Error(t, err)
+	_, err = sc.Shutdown(context.Background(), &sdk.Empty{})
+	assert.NoError(t, err)
+	assertStateChange(v1alpha1.GameServerStateShutdown, 2*time.Second, func(status v1alpha1.GameServerStatus) {
+		assert.Nil(t, status.ReservedUntil)
 	})
+
+	select {
+	case current := <-state:
```

Comment:
With 3 test cases all waiting 4+ seconds, this change adds a considerable amount of time to unit tests (which should always be blazingly fast). It would be nice to see if there is a way to do this that completes nearly instantly instead (e.g. my comment about clock manipulation above).

Reply:
Agreed, there are in parallel though.